src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+
+axiomatization
+where
+  hash_linear_order: "OFCLASS(hash, linorder_class)"
+
+instance hash :: linorder
+  by (fact hash_linear_order)
+
+subsection \<open>Blocks\<close>
+
+datatype block = Block (hash: \<open>hash\<close>)
+
+subsection \<open>Chains\<close>
+
+type_synonym chain = "block list"
+
+type_synonym chain_index = nat
+
+abbreviation (input) tip_index :: "chain \<Rightarrow> chain_index" where
+  "tip_index \<C> \<equiv> length \<C> - 1"
+
+abbreviation (input) tip :: "chain \<Rightarrow> block" where
+  "tip \<C> \<equiv> \<C> ! (tip_index \<C>)"
+
+subsection \<open>Points\<close>
+
+type_synonym point = hash
+
+definition block_point :: "block \<Rightarrow> point" where
+  [simp]: "block_point \<equiv> hash"
+
+subsection \<open>Protocol Parameters\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_chain_update: ChainUpdate
+
+type_synonym consumer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  phase \<comment> \<open>current phase\<close> \<times>
+  (chain \<Rightarrow> point list)" \<comment> \<open>function for selecting candidate points\<close>
+
+type_synonym producer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  chain_index \<comment> \<open>read pointer\<close> \<times>
+  bool \<comment> \<open>whether the next response to \<open>RequestNext\<close> should be \<open>RollBackward\<close>\<close>"
+
+locale chain_sync =
+  fixes initial_params\<^sub>c :: consumer_parameters
+    and initial_params\<^sub>p :: producer_parameters
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Producer |
+  Consumer
+
+subsection \<open>State Machines\<close>
+
+datatype state =
+  Idle |
+  Intersect |
+  CanAwait
+
+datatype message =
+  is_next_request: RequestNext |
+  is_roll_forward: RollForward \<open>block\<close> \<open>block\<close> |
+  is_roll_backward: RollBackward \<open>point\<close> \<open>block\<close> |
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>point list\<close> |
+  is_intersect_found: IntersectFound \<open>point\<close> \<open>block\<close> |
+  is_intersect_not_found: IntersectNotFound \<open>block\<close>
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Consumer" |
+  "agent_in_state' _ = Producer"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (
+    partial_case m of
+      FindIntersect _ \<Rightarrow> Intersect |
+      RequestNext \<Rightarrow> CanAwait)" |
+  "next_state' Intersect m = (
+    partial_case m of
+      IntersectFound _ _ \<Rightarrow> Idle |
+      IntersectNotFound _ \<Rightarrow> Idle)" |
+  "next_state' CanAwait m = (
+    partial_case m of
+      RollForward _ _ \<Rightarrow> Idle |
+      RollBackward _ _ \<Rightarrow> Idle |
+      AwaitReply \<Rightarrow> Idle)" \<comment> \<open>only for this initial implementation\<close>
```

Comment:
I think it’s better to not put the occurrences of `partial_case m of` on lines of their own but instead place them directly after the opening parentheses preceding them. Of course, your current layout better reflects the term structure. However, we actually use `partial_case` terms only because we don’t have direct support for defining partial functions (we don’t have a `partial_primrec`). With such support, we would perform pattern matching for messages on the left-hand sides of the equations and thus on the same lines as the equals signs.

Once the placement of the occurrences of `partial_case m of` has been modified this way, the actual case clauses should be indented one level less, of course.

Also, I strongly advocate for placing the closing parentheses that come after the `partial_case` expressions on lines of their own, aligning them with the beginnings of the equations. After all, they are not logically tied to the last case clauses. Making this change will result in a layout that more resembles the layout of typical `begin`–`end` structures like in [Oberon](https://en.wikipedia.org/wiki/Oberon_(programming_language)).